ros2topic/ros2topic/verb/delay.py

```diff
@@ -0,0 +1,236 @@
+# POSSIBILITY OF SUCH DAMAGE.
+
+# This file is originally from:
+# https://github.com/ros/ros_comm/blob/6e5016f4b2266d8a60c9a1e163c4928b8fc7115e/tools/rostopic/src/rostopic/__init__.py
+
+import importlib
+import math
+
+from time import sleep
+
+import rclpy
+
+from rclpy.expand_topic_name import expand_topic_name
+from rclpy.qos import qos_profile_sensor_data
+from rclpy.validate_full_topic_name import validate_full_topic_name
+from ros2cli.node.direct import DirectNode
+from ros2topic.api import get_topic_names_and_types
+from ros2topic.api import TopicNameCompleter
+from ros2topic.verb import VerbExtension
+
+
+class DelayVerb(VerbExtension):
+    """Display delay of topic from timestamp in header."""
+
+    def add_arguments(self, parser, cli_name):
+        arg = parser.add_argument(
+            'topic',
+            help='Topic name to be calcurated the delay')
+        arg.completer = TopicNameCompleter(
+            include_hidden_topics_key='include_hidden_topics')
+        parser.add_argument(
+            '--window', '-w', default=-1, type=int,
+            help='window size, in # of messages, for calculating rate, '
+                 'string to (default: -1)')
+
+    def main(self, *, args):
+        return main(args)
+
+
+def main(args):
+    with DirectNode(args) as node:
+        _rostopic_delay(node.node, args.topic, window_size=args.window, blocking=True)
+
+
+class ROSTopicDelay(object):
+    """Receives messages for a topic and computes timestamp delay."""
+
+    def __init__(self, node, window_size):
+        import threading
+        self.lock = threading.Lock()
+        self.last_msg_tn = 0
+        self.msg_t0 = -1.
+        self.msg_tn = 0
+        self.delays = []
+
+        # can't have infinite window size due to memory restrictions
+        if window_size < 0:
+            window_size = 50000
+        self.window_size = window_size
+
+        self._clock = node.get_clock()
+
+    def callback_delay(self, msg):
+        """
+        Calculate delay time.
+
+        :param msg: Message instance
+        """
+        if not hasattr(msg, 'header'):
+            raise RuntimeError('msg does not have header')
+        with self.lock:
+            curr_rostime = self._clock.now()
+
+            # time reset
+            if curr_rostime.nanoseconds == 0:
+                if len(self.delays) > 0:
+                    print('time has reset, resetting counters')
+                    self.delays = []
+                return
+
+            curr = curr_rostime.nanoseconds
+            if self.msg_t0 < 0 or self.msg_t0 > curr:
+                self.msg_t0 = curr
+                self.msg_tn = curr
+                self.delays = []
+            else:
+                curr_sec = curr_rostime.to_msg().sec
+                curr_nanosec = curr_rostime.to_msg().nanosec
+
+                delay = ((curr_sec + curr_nanosec * 1e-9) -
+                         (msg.header.stamp.sec + msg.header.stamp.nanosec * 1e-9))
```

Comment:
You should be able to compute the delay (as a Duration) by doing `curr_rostime.to_msg() - rclpy.Time.from_msg(msg.header.stamp)` (see https://github.com/ros2/rclpy/blob/62012d3650de790d89c725232451c4e998b396fb/rclpy/rclpy/time.py#L131) and store the duration instance in `self.delays`.